pkg/docker/docker.go

```diff
@@ -472,17 +480,50 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) (err error) {
 		}
 	}
 
-	glog.V(2).Infof("Waiting for container")
-	exitCode, err := d.client.WaitContainer(container.ID)
-	wg.Wait()
-	if err != nil {
-		return err
-	}
-	glog.V(2).Infof("Container exited")
+	if opts.TargetImage {
 
-	if exitCode != 0 {
-		return errors.NewContainerError(container.Name, exitCode, "")
+		lco := docker.ListContainersOptions{
+			All: true,
+		}
+		conts, lcerr := d.client.ListContainers(lco)
+		var liveports []docker.APIPort
+		if lcerr == nil {
+			for _, cont := range conts {
+				if cont.ID == container.ID {
+					liveports = cont.Ports
+					break
+				}
+			}
+		}
+
+		glog.V(0).Infof("\n\n\n\n\nThe image %s has been started in container %s as a result of the --run=true option.  The container's stdout/stderr will be redirected to this command's glog output to help you validate its behavior.  You can also inspect the container with docker commands if you like.  If the container is set up to stay running, you will have to Ctrl-C to exit this command, which should also stop the container %s.  This particular invocation attempts to run with the port mappings %+v \n\n\n\n\n", image, container.ID, container.ID, liveports)
+
+		signalChan := make(chan os.Signal, 1)
+		cleanupDone := make(chan bool)
+		signal.Notify(signalChan, os.Interrupt)
+		go func() {
+			for range signalChan {
+				glog.V(2).Info("\nReceived an interrupt, stopping services...\n")
+				cleanupDone <- true
+			}
+		}()
+		<-cleanupDone
+	} else {
+		glog.V(2).Infof("Waiting for container")
+		exitCode, err := d.client.WaitContainer(container.ID)
+		glog.V(2).Infof("Container wait returns with %d and %v\n", exitCode, err)
+
+		wg.Wait()
+		if err != nil {
+			return err
+		}
+
+		glog.V(2).Infof("Container exited")
+		if exitCode != 0 {
+			return errors.NewContainerError(container.Name, exitCode, "")
+		}
 	}
+
```

Comment:
I'd suggest splitting this function into smaller parts, it's almost 200 lines long and it's damn hard to follow efficiently, my heads explodes trying to follow the logic in it. Is it possible somehow?